### ISSUE TITLE
Clang cannot detect some unused definitions

### ISSUE THREAD

**Issue by FloridSleeves** [llvm/llvm-project#54820]
Hi,I'm using Clang for detecting unused definitions in programs. However, I find that Clang fails to detect some of them like:
int main()
{
  int ret = 0;
  ret = foo(6);
  return ret;
}
Clang currently does not report the unused `ret=0` and `ret=foo(6)` definitions even with flag '-Wunused'. So I wonder is there any reason for not detecting such unused definitions?

**Comment by pogo59** [comment]
I assume you have something like `int foo(int);` declared before main.
The `ret=foo(6);` is used by `return ret;` so should not be diagnosed.

Arguably the `= 0` could be diagnosed, but there is also a common idiom that requires all variable declarations to have an initializer, so that no uninitialized values can be used within a program.  We might be deliberately not diagnosing such cases.

Curiously, if I change `return ret;` to `return 0;` (so that `ret` is genuinely unused)` I still don't get a diagnostic.  That seems worth investigation by someone familiar with how these diagnostics work.

**Comment by FloridSleeves** [comment]
> I assume you have something like `int foo(int);` declared before main. The `ret=foo(6);` is used by `return ret;` so should not be diagnosed.
> 
> Arguably the `= 0` could be diagnosed, but there is also a common idiom that requires all variable declarations to have an initializer, so that no uninitialized values can be used within a program. We might be deliberately not diagnosing such cases.
> 
> Curiously, if I change `return ret;` to `return 0;` (so that `ret` is genuinely unused)` I still don't get a diagnostic. That seems worth investigation by someone familiar with how these diagnostics work.

Yep that's what I mean. Even the 'ret=foo(6)' cannot be detected when it's unused.

**Comment by FloridSleeves** [comment]
Any comment on this?

**Comment by martong** [comment]
> I think Clang is only able to detect if a definition is not used at all. It sounds like what you are looking for is an analysis that detects a particular assigned value is not read.
> 
> @martong do you know if this is something the static analyzer could detect?

Yes, `deadcode.DeadStores` can detect this case https://godbolt.org/z/n17o4PTG5
```
int foo(int);
int main()
{
  int ret = foo(6);
  return 0;
}

<source>:4:7: warning: Value stored to 'ret' during its initialization is never read [deadcode.DeadStores]
  int ret = foo(6);
      ^~~   ~~~~~~
```
However, (interestingly enough) we don't get the warning for this:
```
int main()
{
  int ret = 0;
  return 0;
}
```

**Comment by martong** [comment]
`int ret = 0;` is handled as a special case and considered as defensive programming. 
```
              // Special case: check for initializations with constants.
              //
              //  e.g. : int x = 0;
              //         struct A = {0, 1};
              //         struct B = {{0}, {1, 2}};
              //
              // If x is EVER assigned a new value later, don't issue
              // a warning.  This is because such initialization can be
              // due to defensive programming.
```
Arguably, we might handle only zero initialization as such, and other constant values should be reported? @haoNoQ, @steakhal, what do you think?
